arangod/VocBase/KeyGenerator.cpp

```diff
@@ -194,76 +179,123 @@ class TraditionalKeyGenerator : public KeyGenerator {
   void toVelocyPack(arangodb::velocypack::Builder& builder) const override {
     KeyGenerator::toVelocyPack(builder);
     builder.add("type", VPackValue("traditional"));
-      
-    MUTEX_LOCKER(mutexLocker, _lock);
-    builder.add(StaticStrings::LastValue, VPackValue(_lastValue));
   }
 
  protected:
+  /// @brief generate a new key value (internal)
   virtual uint64_t generateValue() = 0;
 
- private:
-  mutable arangodb::Mutex _lock;
-
-  uint64_t _lastValue;
+  /// @brief track a value (internal)
+  virtual void track(uint64_t value) = 0;
 };
 
 /// @brief traditional key generator for a single server
 class TraditionalKeyGeneratorSingle final : public TraditionalKeyGenerator {
  public:
   /// @brief create the generator
   explicit TraditionalKeyGeneratorSingle(bool allowUserKeys) 
-    : TraditionalKeyGenerator(allowUserKeys) {}
+    : TraditionalKeyGenerator(allowUserKeys), _lastValue(0) {
+    TRI_ASSERT(!ServerState::instance()->isCoordinator());
+  }
+  
+  /// @brief build a VelocyPack representation of the generator in the builder
+  void toVelocyPack(arangodb::velocypack::Builder& builder) const override {
+    TraditionalKeyGenerator::toVelocyPack(builder);
+     
+    // add our specific stuff 
+    MUTEX_LOCKER(mutexLocker, _lock);
+    builder.add(StaticStrings::LastValue, VPackValue(_lastValue));
+  }
 
  private:
-  /// @brief generate a key
+  /// @brief generate a key value (internal)
   uint64_t generateValue() override {
-    return TRI_NewTickServer();
+    uint64_t tick = TRI_NewTickServer();
+
+    if (ADB_UNLIKELY(tick == UINT64_MAX)) {
+      // out of keys
+      return 0;
+    }
+
+    // keep track of last assigned value, and make sure the value
+    // we hand out is always higher than it
+    {
+      MUTEX_LOCKER(mutexLocker, _lock);
+      
+      if (ADB_UNLIKELY(_lastValue >= UINT64_MAX - 1ULL)) {
+        // oops, out of keys!
+        return 0;
+      }
+
+      if (tick <= _lastValue) {
+        tick = ++_lastValue;
+      } else {
+        _lastValue = tick;
+      }
+    }
+
+    return tick;
+  }
+
+  /// @brief track a key value (internal)
+  void track(uint64_t value) override {
+    MUTEX_LOCKER(mutexLocker, _lock);
+
+    if (value > _lastValue) {
+      // and update our last value
+      _lastValue = value;
+    }
   }
+
+ private:
+  mutable arangodb::Mutex _lock;
+
+  uint64_t _lastValue;
 };
 
 /// @brief traditional key generator for a coordinator
+/// please note that coordinator-based key generators are frequently
+/// created and discarded, so ctor & dtor need to be very efficient.
+/// additionally, do not put any state into this object, as for the
+/// same logical collection the ClusterInfo may create many different 
+/// temporary LogicalCollection objects one after the other, which
+/// will also discard the collection's particular KeyGenerator object!
 class TraditionalKeyGeneratorCluster final : public TraditionalKeyGenerator {
```

Comment:
could we not share the key generator objects between LogicalCollection instances ?

Reply:
we should rather use `LogicalCollection` objects instead of just `KeyGenerator` objects.